### PR TITLE
tests(migration): typescript

### PR DESCRIPTION
**Pull request by dhruvdutt**
**What kind of change does this PR introduce?**
Tests: Refactoring to TypeScript

**Did you add tests for your changes?**
Yes

**If relevant, did you update the documentation?**
NA

**Summary**
Added `ts-test` branch on upstream. We will keep working on this branch to migrate all tests to TypeScript.

**Does this PR introduce a breaking change?**
No

**Other information**
cc @sendilkumarn @ematipico 

Packages that contains tests to port to typescript
- [x] add
- [x] generate-loader
- [x] generate-plugin
- [x] generators
- [X] info
- [x] init
- [x] make
- [X] migrate
- [x] remove
- [x] serve
- [x] update
- [x] utils
- [x] webpack-scaffold: #686 by @hemal7735

### CODE REVIEW

**Comment by ematipico** [comment]
Started the migration of other files. Please follow the same pattern

**Comment by rishabh3112** [comment]
Can I help in this migration too?
If so can I get pointers from where to get started 😅

**Comment by ematipico** [comment]
Hi @rishabh3112, surely you can!
We are migrating the tests inside the different packages from JavaScript to TypeScript.

There's a task list that I added inside the main issue with the different packages and we are going to tick the ones that are migrated. 

Please feel free to pick one and tell us, so we are going to update the main thread and we are not going to work on the same thing.

Most of them don't require an actual tests because they are just wrappers (add, update, etc.).

The main bulk will be `utils` package. So I'd suggest to take your time with one/two tests files and start the migration.

Please use the `ts-test` branch to commit your changes

**Comment by rishabh3112** [comment]
Thanks @ematipico 🙌,
I will look into this and let everyone know on which package I would be working.
(Sorry for late reply though 😅)

**Comment by rishabh3112** [comment]
I am facing a little difficulty in setting up.
How can I keep intact my npm installation intact while working over repo?
@ematipico, @dhruvdutt, @ev1stensberg?

**Comment by dhruvdutt** [comment]
Since the new infra is now ready, we can start working on this. Every contribution is welcomed. We will continue working on `ts-test` branch.

@rishabh3112 Not sure why GitHub isn't allowing me to assign you. 😅

**Comment by rishabh3112** [comment]
> Since the new infra is now ready, we can start working on this. Every contribution is welcomed. We will continue working on `ts-test` branch.
> 
> @rishabh3112 Not sure why GitHub isn't allowing me to assign you.

No problem @dhruvdutt :)

**Comment by dhruvdutt** [comment]
@hemal7735 You can start off with smaller packages that have fewer tests. Please just comment here or something before you pick a package, so there's clarity.
Also, please make sure you do your PRs on `ts-test` and a separate PR for every package.

**Comment by hemal7735** [comment]
I'm starting off with `webpack-scaffold`.

**Comment by ematipico** [comment]
The test is passing a function. As you can the utility should create a function, right?

**Comment by hemal7735** [comment]
utility is creating a function out of the value that is passed. Should not it be `string` instead of `function` ? `createAssetFilterFunction` and `createExternalFunction` also expects `string`.
This might get missed because we were not using TS.

**Comment by hemal7735** [comment]
`defineTest` has last params `object` and `string` - https://github.com/webpack/webpack-cli/blob/ts-test/packages/utils/defineTest.ts#L121-L122
but recursive-parser.test.js is passing wrongly. https://github.com/webpack/webpack-cli/blob/ts-test/packages/utils/recursive-parser.test.js#L5

Are test cases broken or it is intentional?

**Comment by ematipico** [comment]
@dhruvdutt could you please help @hemal7735 ? I know you worked on that

**Comment by rishabh3112** [comment]
I am working on `generator-loader` package.
what are these `.tpl` files in there?
And also many packages don't have test, they should be removed from checklist.

**Comment by ematipico** [comment]
They are template files to generate the scaffolding of a plugin or loader. If files don't have test that's fine. We just plan to migrate the existing ones

**Comment by rishabh3112** [comment]
Why tpl files are under test folder @ematipico

**Comment by hemal7735** [comment]
I was working on `utils` package. We have the following tests. The ticked ones are done.
I need help with the unticked one. 

- [x] ast-utils.test.ts           
- [x] npm-exists.test.ts          
- [x] package-manager.test.ts     
- [x] resolve-packages.test.ts
- [x] is-local-path.test.ts       
- [ ] npm-packages-exists.test.ts - **TS mock issue - need help here** 
- [ ] recursive-parser.test.tx -- `defineTes`t parameter issue  , already mentioned above
- [x] validate-identifier.test.ts

**Comment by ematipico** [comment]
Tpl files are under test folder because it generates the files for testing as well

**Comment by ematipico** [comment]
@hemal7735 feel free to make a pr against this brach so we can help you :)

**Comment by hemal7735** [comment]
@evenstensberg @ematipico I'm not sure where are we on this?

**Comment by dhruvdutt** [comment]
The branch is up-to-date with master. Merging is blocked because "1 review requesting changes by reviewers with write access."
@evenstensberg I will prioritize this PR and wrap it up in a week.

**Comment by ematipico** [comment]
Looks good to me

**Comment by dhruvdutt** [comment]
@evenstensberg Can you please take a look at this?

**Comment by ematipico** [comment]
@evenstensberg PR is ready to be merged. Could you please leave your blessing? 😃

File: packages/migrate/migrate.ts [hunk 1]
```diff
@@ -42,7 +42,7 @@ const transformsObject: ITransformsObject = {
 };
 /* tslint:enable object-literal-sort-keys */
 
-export const transformations =
+export const transformations: any =
```
Comment (evenstensberg) [inline packages/migrate/migrate.ts]: interface this

File: packages/migrate/namedModulesPlugin/__tests__/namedModulesPlugin.test.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import defineTest from "@webpack-cli/utils/defineTest";
+import { join } from "path";
+defineTest(join(__dirname, ".."), "namedModulesPlugin", "namedModulesPlugin-0");
```
Comment (evenstensberg) [inline packages/migrate/namedModulesPlugin/__tests__/namedModulesPlugin.test.ts]: path into variable

File: packages/migrate/outputPath/__tests__/outputPath.test.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import defineTest from "@webpack-cli/utils/defineTest";
+import { join } from "path";
+
+defineTest(join(__dirname, ".."), "outputPath", "outputPath-0");
```
Comment (evenstensberg) [inline packages/migrate/outputPath/__tests__/outputPath.test.ts]: in a variable

File: packages/migrate/removeDeprecatedPlugins/__tests__/removeDeprecatedPlugins.test.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import defineTest from "@webpack-cli/utils/defineTest";
+import { join } from "path";
+
+defineTest(
```
Comment (evenstensberg) [inline packages/migrate/removeDeprecatedPlugins/__tests__/removeDeprecatedPlugins.test.ts]: in variable

File: packages/migrate/removeJsonLoader/__tests__/removeJsonLoader.test.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import defineTest from "@webpack-cli/utils/defineTest";
+import * as path from "path";
+defineTest(
+  path.join(__dirname, ".."),
```
Comment (evenstensberg) [inline packages/migrate/removeJsonLoader/__tests__/removeJsonLoader.test.ts]: variable

File: packages/migrate/resolve/__tests__/__testfixtures__/resolve.input.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+const path = require("patch");
+module.exports = [
```
Comment (evenstensberg) [inline packages/migrate/resolve/__tests__/__testfixtures__/resolve.input.js]: `{resolve}`

File: packages/migrate/resolve/__tests__/resolve.test.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import defineTest from "@webpack-cli/utils/defineTest";
+import * as path from "path";
+
+defineTest(path.join(__dirname, ".."), "resolve");
```
Comment (evenstensberg) [inline packages/migrate/resolve/__tests__/resolve.test.ts]: variable

File: packages/migrate/uglifyJsPlugin/__tests__/uglifyJsPlugin.test.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import defineTest from "@webpack-cli/utils/defineTest";
+import { join } from "path";
+defineTest(join(__dirname, ".."), "uglifyJsPlugin", "uglifyJsPlugin-0");
```
Comment (evenstensberg) [inline packages/migrate/uglifyJsPlugin/__tests__/uglifyJsPlugin.test.ts]: variable

File: packages/utils/__tests__/ast-utils.test.ts [hunk 9]
```diff
@@ -191,8 +192,8 @@ const a = { plugs: [] }
 			expect(
 				ast
 					.find(j.ObjectExpression)
-					.filter(p => utils.findObjWithOneOfKeys(p, ["a"]))
-					.size()
+					.filter((p) => utils.findObjWithOneOfKeys(p, ["a"]))
```
Comment (evenstensberg) [inline packages/utils/__tests__/ast-utils.test.ts]: types here

File: packages/utils/__tests__/ast-utils.test.ts [hunk 10]
```diff
@@ -299,20 +300,20 @@ const a = { plugs: [] }
...
+					},
 				],
-				nice: "':)'",
-				man: "() => duper"
 			};
 
 			const root = ast.find(j.ObjectExpression);
 
-			utils.findRootNodesByName(j, root, "entry").forEach(p => {
+			utils.findRootNodesByName(j, root, "entry").forEach((p) => {
```
Comment (evenstensberg) [inline packages/utils/__tests__/ast-utils.test.ts]: types here

File: packages/utils/__tests__/recursive-parser.test.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
+"use strict";
+
+import { join } from "path";
+import defineTest from "../defineTest";
+
+defineTest(join(__dirname, ".."), "init", "fixture-1", "entry", {
```
Comment (evenstensberg) [inline packages/utils/__tests__/recursive-parser.test.ts]: variable

File: packages/utils/__tests__/resolve-packages.test.ts [hunk 12]
```diff
@@ -17,17 +17,15 @@ function spawnChild(pkg) {
 }
 
 function getLoc(option) {
-	let packageModule = [];
-	option.filter(pkg => {
-		mockPromise(spawnChild(pkg)).then(_ => {
+	const packageModule = [];
+	option.filter((pkg) => {
+		mockPromise(spawnChild(pkg)).then((_) => {
```
Comment (evenstensberg) [inline packages/utils/__tests__/resolve-packages.test.ts]: convention?

Comment (ematipico) [reply packages/utils/__tests__/resolve-packages.test.ts]: This is for later

File: packages/migrate/namedModulesPlugin/__tests__/namedModulesPlugin.test.ts (hunk 2 above)
Comment (ematipico) [reply packages/migrate/namedModulesPlugin/__tests__/namedModulesPlugin.test.ts]: Is it trivial?

Comment (evenstensberg) [reply packages/migrate/namedModulesPlugin/__tests__/namedModulesPlugin.test.ts]: doesn't look readable for people maintaining this

**Comment by misterdev** [comment]
Can I help here?

**Comment by dhruvdutt** [comment]
@misterdev Sure, thing. You can start going through review comments and make PR on this as a base branch.

**Comment by misterdev** [comment]
@dhruvdutt @evenstensberg I've worked on the pending requested changes and made a PR, would be great to have any kind of feedback  👍

**Comment by dhruvdutt** [comment]
Working on rebasing this.

**Comment by evenstensberg** [comment]
Status?

**Comment by sendilkumarn** [comment]
I merged upstream changes. Hopefully, this one will be good to go 🤞

File: package.json [hunk 13]
```diff
@@ -43,7 +43,8 @@
     "travis:integration": "npm run build && npm run test && npm run reportCoverage",
     "travis:lint": "npm run build && npm run lint && npm run tslint",
     "tslint": "tslint -c tslint.json \"packages/**/*.ts\"",
-    "watch": "npm run build && tsc -w"
+    "watch": "npm run build && tsc -w",
+    "test:local": "jest --watch"
```
Comment (evenstensberg) [inline package.json]: could we remove this?

Comment (sendilkumarn) [reply package.json]: makes sense 👍

**Comment by evenstensberg** [comment]
As a final check, could you pull master branch, check if we are missing something?

**Comment by sendilkumarn** [comment]
This has latest master in it. so we can merge

**Comment by sendilkumarn** [comment]
@evenstensberg Let us not wait further. I will merge this. I think on the Typescript side it is better to follow some conventions like utilizing type inferences and being strict on any (things like that). I will create a new PR out of this.

**Pull request merged**